alibi/explainers/similarity/backends/pytorch/base.py

```diff
@@ -0,0 +1,83 @@
+"""`pytorch` backend for similarity explainers.
+
+Methods unique to the `pytorch` backend are defined here. The interface this class defines syncs with the `tensorflow`
+backend in order to ensure that the similarity methods only require to match this interface.
+"""
+
+from typing import Callable, Union, Optional
+
+import numpy as np
+import torch.nn as nn
+import torch
+
+
+class _PytorchBackend:
+    device: Optional[torch.device] = None  # device used by `pytorch` backend
+
+    @staticmethod
+    def get_grads(
+            model: nn.Module,
+            X: torch.Tensor,
+            Y: torch.Tensor,
+            loss_fn: Callable[[torch.Tensor, torch.Tensor], torch.Tensor],
+    ) -> np.ndarray:
+        """
+        Computes the gradients of the loss function with respect to the model's parameters for a single training and
+        target pair.
+
+        Parameters:
+        -----------
+        model:
+            The model to compute gradients for.
+        X:
+            The input data.
+        Y:
+            The target data.
+        loss_fn:
+            The loss function to use.
+
+        Returns:
+        --------
+        grads:
+            The gradients of the loss function with respect to the model's parameters. This is returned as a flattened \
+            array.
+        """
+
+        model.zero_grad()
+        initial_model_state = model.training
+        model.train(False)
+        output = model(X)
+        loss = loss_fn(output, Y)
+        loss.backward()
+        model.train(initial_model_state)
+        return np.concatenate([_PytorchBackend.to_numpy(param.grad).reshape(-1)
+                               for param in model.parameters()])
+
+    @staticmethod
+    def to_tensor(X: np.ndarray) -> torch.Tensor:
```

Comment:
Not sure if needed, but the params and return description is missing for the majority of methods in this module.